camellia/src/camellia.rs

```diff
@@ -0,0 +1,352 @@
+                f.write_str(concat!(stringify!($name), " { ... }"))
+            }
+        }
+
+        impl AlgorithmName for $name {
+            fn write_alg_name(f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                f.write_str(concat!(stringify!($name)))
+            }
+        }
+
+        #[cfg(feature = "zeroize")]
+        #[cfg_attr(docsrs, doc(cfg(feature = "zeroize")))]
+        impl Drop for $name {
+            fn drop(&mut self){
+                self.kw.zeroize();
+                self.k.zeroize();
+                self.ke.zeroize();
+            }
+        }
+
+        #[cfg(feature = "zeroize")]
+        #[cfg_attr(docsrs, doc(cfg(feature = "zeroize")))]
+        impl ZeroizeOnDrop for $name {}
+
+        cipher::impl_simple_block_encdec!(
+            $name, U16, cipher, block,
+            encrypt: {
+                let b = block.get_in();
+                let mut d1 = u64::from_be_bytes(b[0..8].try_into().unwrap());
+                let mut d2 = u64::from_be_bytes(b[8..16].try_into().unwrap());
+
+                d1 ^= cipher.kw[0];
+                d2 ^= cipher.kw[1];
+
+                d2 ^= f(d1, cipher.k[0]);
+                d1 ^= f(d2, cipher.k[1]);
+                for round in (2..$rounds).step_by(2) {
+                    if round % 6 == 0 {
+                        d1 = fl(d1, cipher.ke[round / 3 - 2]);
+                        d2 = flinv(d2, cipher.ke[round / 3 - 1]);
+                    }
+                    d2 ^= f(d1, cipher.k[round]);
+                    d1 ^= f(d2, cipher.k[round + 1]);
+                }
+
+                d2 ^= cipher.kw[2];
+                d1 ^= cipher.kw[3];
+
+                BE::write_u64_into(&[d2, d1], block.get_out());
+            }
+            decrypt: {
+                let b = block.get_in();
+                let mut d1 = u64::from_be_bytes(b[0..8].try_into().unwrap());
+                let mut d2 = u64::from_be_bytes(b[8..16].try_into().unwrap());
+
+                d2 ^= cipher.kw[3];
+                d1 ^= cipher.kw[2];
+
+                d2 ^= f(d1, cipher.k[$rounds - 1]);
+                d1 ^= f(d2, cipher.k[$rounds - 2]);
+                for round in (0..$rounds - 2).rev().step_by(2) {
+                    if (round + 1) % 6 == 0 {
+                        d1 = fl(d1, cipher.ke[round / 3]);
+                        d2 = flinv(d2, cipher.ke[round / 3 - 1]);
+                    }
+                    d2 ^= f(d1, cipher.k[round]);
+                    d1 ^= f(d2, cipher.k[round - 1]);
+                }
+
+                d1 ^= cipher.kw[1];
+                d2 ^= cipher.kw[0];
+
+                BE::write_u64_into(&[d2, d1], block.get_out());
+            }
+        );
+    };
+}
+
+impl_camellia!(Camellia128, 18, U16, "Camellia-128 block cipher instance.");
+impl_camellia!(Camellia192, 24, U24, "Camellia-192 block cipher instance.");
+impl_camellia!(Camellia256, 24, U32, "Camellia-256 block cipher instance.");
```

Comment:
It seems like this could be impl'd in terms of a generic core type with (const) generic parameters and type aliases for the specific key sizes, rather than using macros.

Likewise it seems like `gen_subkeys!` could potentially be a function (or rather, two functions) rather than macros.

Reply:
`gen_subkeys!` has been fixed.